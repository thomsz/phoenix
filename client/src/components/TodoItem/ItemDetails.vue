<template>
  <div class="p-4">
    <EditInPlace
      name="description"
      type="textarea"
      class="min-h-[2.5rem]"
      v-model="form.description"
    >
      <span v-if="form.description">{{ form.description }}</span>
      <span
        v-else
        class="text-slate-200"
      >
        Add description
      </span>
    </EditInPlace>
    <hr />
    <EditInPlace
      name="due_date"
      type="date"
      class="min-h-[2.5rem]"
      v-model="form.dueDate"
    >
      <span v-if="form.dueDate">{{ formattedLocaleDueDate }}</span>
      <span
        v-else
        class="text-slate-200"
      >
        Add due date
      </span>
    </EditInPlace>
  </div>
</template>

<script lang="ts">
import { defineComponent } from 'vue'

import EditInPlace from '@/components/common/EditInPlace.vue'

import type Todo from '@/types/Todo'
import type { PropType } from 'vue'

export default defineComponent({
  name: 'ItemDetails',

  components: {
    EditInPlace
  },

  props: {
    modelValue: {
      type: Object as PropType<Todo>,
      required: true
    }
  },

  data () {
    return {
      form: this.modelValue
    }
  },

  computed: {
    formattedLocaleDueDate (): string | undefined {
      return new Date(this.form.dueDate).toLocaleDateString()
    }
  },

  watch: {
    form: {
      deep: true,
      handler (form: Todo) {
        this.$emit('update:modelValue', form)
      }
    }
  }
})
</script>
