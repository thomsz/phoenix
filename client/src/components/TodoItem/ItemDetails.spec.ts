import factory from '@/tests/services/factory'
import { faker } from '@faker-js/faker'
import { mount, shallowMount } from '@vue/test-utils'

import ItemDetails from '@/components/TodoItem/ItemDetails.vue'

import type Todo from '@/types/Todo'


describe('ItemDetails', () => {
  const fakeItem: Todo = {
    ...factory.generate('Todo'),
    dueDate: '2022-08-20T10:15:59.765Z',
    description: 'Fake description'
  }
  
  const fakeItemWithMissingData: Todo = {
    ...fakeItem,
    dueDate: '',
    description: ''
  }

  ;[
    ['with details', fakeItem],
    ['with missing details', fakeItemWithMissingData]
  ].forEach(([descriptor, item]) => {
    it(`should render details: ${descriptor}`, () => {
      const wrapper = mount(ItemDetails, {
        props: {
          modelValue: item
        }
      })
      
      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  it('should update model value when data changes', async () => {
    const wrapper = shallowMount(ItemDetails, {
      props: {
        modelValue: factory.generate('Todo')
      }
    })

    const updatedItem = {
      ...fakeItem,
      title: faker.random.words()
    }

    await wrapper.setData({ form: updatedItem })
    expect(wrapper.emitted('update:modelValue')?.[0][0]).toEqual(updatedItem)
  })
})
