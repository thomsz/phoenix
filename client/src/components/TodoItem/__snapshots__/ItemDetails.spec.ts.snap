// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemDetails should render details: with details 1`] = `
"<div class=\\"p-4\\">
  <div class=\\"edit-in-place min-h-[2.5rem]\\">
    <div class=\\"p-2 text-slate-500 whitespace-pre-wrap\\" data-testid=\\"interactive-element\\"><span>Fake description</span></div>
  </div>
  <hr>
  <div class=\\"edit-in-place min-h-[2.5rem]\\">
    <div class=\\"p-2 text-slate-500\\" data-testid=\\"interactive-element\\"><span>8/20/2022</span></div>
  </div>
</div>"
`;

exports[`ItemDetails should render details: with missing details 1`] = `
"<div class=\\"p-4\\">
  <div class=\\"edit-in-place min-h-[2.5rem]\\">
    <div class=\\"p-2 text-slate-500 whitespace-pre-wrap\\" data-testid=\\"interactive-element\\"><span class=\\"text-slate-200\\"> Add description </span></div>
  </div>
  <hr>
  <div class=\\"edit-in-place min-h-[2.5rem]\\">
    <div class=\\"p-2 text-slate-500\\" data-testid=\\"interactive-element\\"><span class=\\"text-slate-200\\"> Add due date </span></div>
  </div>
</div>"
`;
